### DISCUSSION TITLE
File attachments in multipart HTTP requests

### DISCUSSION THREAD

**Discussion by richard-coles** [nextgenhealthcare/connect#4871] [General]
I'm having a bit of trouble when I'm posting files via HTTP into a channel and I'm running out of hair to pull out. I'm sure there's an obvious answer to this but can't work out what it is.

I have two channels:
The first is setup with an HTTP Receiver and I've selected _Message Content: XML Body_. With this when I send a file to it through Postman I can see the request is converted to XML and the file content is Base64 encoded inside the `<content>` tag. I can then spit the file out to disk via a File Writer destination connector and all is fine, the file is an exact copy of the original.

My second channel is also setup as a HTTP Receiver but with _Message Content: Plain Body_. However I've put the following code  in the PreProcessor script...
```
var headers = sourceMap.get('headers');
message = HTTPUtil.httpBodyToXml(message, headers.getHeader("content-type"));
return message;
```
To the untrained eye it looks like that code achieves the same result as selecting _Message Content: XML Body_ in the source connector. I can see the request converted to XML with the file content Base64 encoded within the `<content>` tag. And the channel still executes fine spitting a file out to disk. Except when you look closely at the Base64 string it has various differences throughout, resulting in the file being corrupted when it's written out again.

I'm assuming it's something to do with encoding but I'm at a loss as to why it's happening or how to resolve it. Can anyone help or give some pointers? Or is there another way of nicely handling the message payload in it's raw form without converting to XML?

**Comment by pacmano1** [comment]
Maybe post both channels so people can replicate your problem.

**Reply by richard-coles** [reply]
Sure, here's a channel group with both channels in - they're as basic as you can get, the only real code in them is what I've written in the original post. Note they assume it's a PDF file you're uploading so attaches a ".pdf" extension to the file when it's written out.

Also attached a postman project to send a file to the channels as part of a multipart HTTP request. Again it's just the very basics, not doing anything clever. I just switch the port number in the URL to point it at the two different channels

(attachments have been renamed to .txt as it won't allow me to upload XML or JSON files)

[Multipart HTTP Request File Test.postman_collection.json.txt](https://github.com/nextgenhealthcare/connect/files/7624720/Multipart.HTTP.Request.File.Test.postman_collection.json.txt)

[Multipart HTTP Request File Test.xml.txt](https://github.com/nextgenhealthcare/connect/files/7624652/Multipart.HTTP.Request.File.Test.xml.txt)
